lib/spack/spack/build_environment.py

```diff
@@ -865,39 +799,173 @@ def setup_package(pkg, dirty, context='build'):
-    pkg = spec.package
-
-    # Maps the context to deptype and method to be called
-    deptype_and_method = {
-        'build': (('build', 'link', 'test'),
-                  'setup_dependent_build_environment'),
-        'run': (('link', 'run'), 'setup_dependent_run_environment'),
-        'test': (('link', 'run', 'test'), 'setup_dependent_run_environment')
-    }
-    deptype, method = deptype_and_method[context]
-
-    root = context == 'test'
-    for dspec in spec.traverse(order='post', root=root, deptype=deptype):
-        dpkg = dspec.package
-        set_module_variables_for_package(dpkg)
-        # Allow dependencies to modify the module
-        dpkg.setup_dependent_package(pkg.module, spec)
-        getattr(dpkg, method)(env, spec)
+
+    # Note: see computation of 'custom_mod_deps' and 'exe_deps' later in this
+    # function; these sets form the building blocks of those collections.
+    build_deps      = set(spec.dependencies(deptype=('build', 'test')))
+    link_deps       = set(spec.traverse(root=False, deptype='link'))
+    build_link_deps = build_deps | link_deps
+    build_and_supporting_deps = set()
+    for build_dep in build_deps:
+        build_and_supporting_deps.update(build_dep.traverse(deptype='run'))
+    run_and_supporting_deps = set(
+        spec.traverse(root=False, deptype=('run', 'link')))
+    test_and_supporting_deps = set()
+    for test_dep in set(spec.dependencies(deptype='test')):
+        test_and_supporting_deps.update(test_dep.traverse(deptype='run'))
+
+    # All dependencies that might have environment modifications to apply
+    custom_mod_deps = set()
+    if context == 'build':
+        custom_mod_deps.update(build_and_supporting_deps)
+        # Tests may be performed after build
+        custom_mod_deps.update(test_and_supporting_deps)
+    else:
+        # test/run context
+        custom_mod_deps.update(run_and_supporting_deps)
+        if context == 'test':
+            custom_mod_deps.update(test_and_supporting_deps)
+    custom_mod_deps.update(link_deps)
+
+    # Determine 'exe_deps': the set of packages with binaries we want to use
+    if context == 'build':
+        exe_deps = build_and_supporting_deps | test_and_supporting_deps
+    elif context == 'run':
+        exe_deps = set(spec.traverse(deptype='run'))
+    elif context == 'test':
+        exe_deps = test_and_supporting_deps
+
+    def default_modifications_for_dep(dep):
+        if (dep in build_link_deps and
+                not is_system_path(dep.prefix) and
+                context == 'build'):
+            prefix = dep.prefix
+
+            env.prepend_path('CMAKE_PREFIX_PATH', prefix)
+
+            for directory in ('lib', 'lib64', 'share'):
+                pcdir = os.path.join(prefix, directory, 'pkgconfig')
+                if os.path.isdir(pcdir):
+                    env.prepend_path('PKG_CONFIG_PATH', pcdir)
+
+        if dep in exe_deps and not is_system_path(dep.prefix):
+            _make_runnable(dep, env)
```

Comment:
Is this going to work for system externals since we clean the PATH? I think if you have an external `cmake` in `/usr` (executable at `/usr/bin/cmake`), then I don't think cmake will be runnable in the build environment with this change.

Reply:
This is supposed to match prior behavior: we didn't add system paths to `PATH`. That has probably worked since we don't clean `PATH`:

```
build_and_supporting_prefixes = filter_system_paths(x.prefix for x in build_and_supporting_deps)

for prefix in reversed(build_and_supporting_prefixes):
  for dirname in ['bin', 'bin64']:
    bin_dir = os.path.join(prefix, dirname)
    if os.path.isdir(bin_dir):
      env.prepend_path('PATH', bin_dir)
```

If we do clean `PATH`, then I think you're right that we'd have to alter the original behavior.

Reply:
Whoa I was so sure we cleaned `PATH`. Shouldn't we be cleaning `PATH`? That's such an obvious way something could sneak into our builds without us knowing.

Reply:
I'm pretty sure that's how we make use of a system-installed `git`/`curl` (i.e. we assume they are in `/usr/bin` and that `/usr/bin` is in `PATH`)